File: web/src/crypto/kdf.ts
/** Master password derivation (Argon2id) and subkeys (crypto_kdf), docs/crypto.md §4 and §5.5. */
import { utf8Encode } from "./encoding";
import { CryptoError } from "./errors";
import { lib } from "./sodium";

export const KEY_BYTES = 32;
export const SALT_BYTES = 16;
export const MIN_PASSWORD_CHARS = 12;
export const MAX_PASSWORD_BYTES = 1024;

export const CTX_AUTH = "srn-auth";
export const CTX_WRAP = "srn-wrap";
export const CTX_RECOVERY_AUTH = "srn-rcva";
export const CTX_RECOVERY_WRAP = "srn-rcvw";
export const CTX_RECOVERY_CHECK = "srn-rcvc";
export const CTX_SEAL = "srn-seal";
export const CTX_EXPORT = "srn-expt";

export interface KdfParams {
  memlimit: number;
  opslimit: number;
}

export const DEFAULT_PARAMS: KdfParams = { memlimit: 64 * 1024 * 1024, opslimit: 3 };
export const FLOOR: KdfParams = DEFAULT_PARAMS;

/** Refuse parameters below the floor, even if the server sends them. */
export function checkParams(params: KdfParams): void {
  if (params.memlimit < FLOOR.memlimit || params.opslimit < FLOOR.opslimit) {
    throw new CryptoError("Argon2id parameters below the floor");
  }
}

export function normalizePassword(password: string): Uint8Array {
  const normalized = password.normalize("NFKC");
  // Count code points, like Python len(): Array.from iterates by code point.
  if (Array.from(normalized).length < MIN_PASSWORD_CHARS) {
    throw new CryptoError(
      `master password must have at least ${String(MIN_PASSWORD_CHARS)} characters`,
    );
  }
  const data = utf8Encode(normalized);
  if (data.length > MAX_PASSWORD_BYTES) {
    throw new CryptoError(`master password must be at most ${String(MAX_PASSWORD_BYTES)} bytes`);
  }
  return data;
}

/** MK = Argon2id(NFKC(password), salt). Never stored, never sent. */
export function deriveMasterKey(
  password: string,
  salt: Uint8Array,
  params: KdfParams = DEFAULT_PARAMS,
): Uint8Array {
  checkParams(params);
  if (salt.length !== SALT_BYTES) throw new CryptoError("salt must be 16 bytes");
  const s = lib();
  const data = normalizePassword(password);
  try {
    return s.crypto_pwhash(
      KEY_BYTES,
      data,
      salt,
      params.opslimit,
      params.memlimit,
      s.crypto_pwhash_ALG_ARGON2ID13,
    );
  } finally {
    s.memzero(data);
  }
}

export function deriveSubkey(
  key: Uint8Array,
  subkeyId: number | bigint,
  context: string,
): Uint8Array {
  if (context.length !== 8 || !/^[\x20-\x7e]{8}$/.test(context)) {
    throw new CryptoError("KDF context must be exactly 8 ASCII characters");
  }
  if (key.length !== KEY_BYTES) throw new CryptoError("KDF key must be 32 bytes");
  const id = BigInt(subkeyId);
  if (id < 0n || id >= 2n ** 64n) throw new CryptoError("invalid subkey id");
  // libsodium.js requires a bigint above 2^32; a bigint is accepted for every id.
  return lib().crypto_kdf_derive_from_key(KEY_BYTES, id, context, key);
}

/** AuthKey (sent to the server) and MEK (kept) from MK. */
export function deriveLoginKeys(masterKey: Uint8Array): {
  authKey: Uint8Array;
  wrapKey: Uint8Array;
} {
  return {
    authKey: deriveSubkey(masterKey, 1, CTX_AUTH),
    wrapKey: deriveSubkey(masterKey, 1, CTX_WRAP),
  };
}

File: api/serenity/crypto/kdf.py
"""Master password derivation (Argon2id) and subkeys (crypto_kdf), docs/crypto.md §4 and §5.5."""

import unicodedata
from dataclasses import dataclass

import nacl.bindings as sodium
import nacl.pwhash

from serenity.crypto.errors import CryptoError

KEY_BYTES = 32
SALT_BYTES = 16
MIN_PASSWORD_CHARS = 12
MAX_PASSWORD_BYTES = 1024
KDF_CONTEXT_BYTES = 8

# Subkey contexts (exactly 8 ASCII characters).
CTX_AUTH = "srn-auth"
CTX_WRAP = "srn-wrap"
CTX_RECOVERY_AUTH = "srn-rcva"
CTX_RECOVERY_WRAP = "srn-rcvw"
CTX_RECOVERY_CHECK = "srn-rcvc"
CTX_SEAL = "srn-seal"
CTX_EXPORT = "srn-expt"


@dataclass(frozen=True)
class KdfParams:
    """Argon2id cost. Values below the floor are always refused."""

    memlimit: int = 64 * 1024 * 1024
    opslimit: int = 3


DEFAULT_PARAMS = KdfParams()
FLOOR = KdfParams()


def check_params(params: KdfParams) -> None:
    if params.memlimit < FLOOR.memlimit or params.opslimit < FLOOR.opslimit:
        raise CryptoError("Argon2id parameters below the floor")


def normalize_password(password: str) -> bytes:
    normalized = unicodedata.normalize("NFKC", password)
    if len(normalized) < MIN_PASSWORD_CHARS:
        raise CryptoError(f"master password must have at least {MIN_PASSWORD_CHARS} characters")
    data = normalized.encode("utf-8")
    if len(data) > MAX_PASSWORD_BYTES:
        raise CryptoError(f"master password must be at most {MAX_PASSWORD_BYTES} bytes")
    return data


def derive_master_key(password: str, salt: bytes, params: KdfParams = DEFAULT_PARAMS) -> bytes:
    """MK = Argon2id(NFKC(password), salt). Never stored, never sent."""
    check_params(params)
    if len(salt) != SALT_BYTES:
        raise CryptoError("salt must be 16 bytes")
    return nacl.pwhash.argon2id.kdf(
        KEY_BYTES,
        normalize_password(password),
        salt,
        opslimit=params.opslimit,
        memlimit=params.memlimit,
    )


def derive_subkey(key: bytes, subkey_id: int, context: str) -> bytes:
    """crypto_kdf_derive_from_key(32, subkey_id, context, key).

    PyNaCl does not expose crypto_kdf; this is libsodium's own construction
    (BLAKE2b, salt = LE64(id) || 0^8, personal = context || 0^8), checked by the vectors.
    """
    ctx = context.encode("ascii")
    if len(ctx) != KDF_CONTEXT_BYTES:
        raise CryptoError("KDF context must be exactly 8 bytes")
    if len(key) != KEY_BYTES:
        raise CryptoError("KDF key must be 32 bytes")
    if not 0 <= subkey_id < 2**64:
        raise CryptoError("invalid subkey id")
    return sodium.crypto_generichash_blake2b_salt_personal(
        b"",
        digest_size=KEY_BYTES,
        key=key,
        salt=subkey_id.to_bytes(8, "little") + bytes(8),
        person=ctx + bytes(8),
    )


def derive_login_keys(master_key: bytes) -> tuple[bytes, bytes]:
    """(AuthKey, MEK) from MK."""
    return derive_subkey(master_key, 1, CTX_AUTH), derive_subkey(master_key, 1, CTX_WRAP)
